encoding/codecv7.go

```diff
@@ -0,0 +1,319 @@
+
+	"github.com/scroll-tech/da-codec/encoding/zstd"
+)
+
+type DACodecV7 struct{}
+
+// Version returns the codec version.
+func (d *DACodecV7) Version() CodecVersion {
+	return CodecV7
+}
+
+// MaxNumChunksPerBatch returns the maximum number of chunks per batch.
+func (d *DACodecV7) MaxNumChunksPerBatch() int {
+	return math.MaxInt
+}
+
+// NewDABlock creates a new DABlock from the given Block and the total number of L1 messages popped before.
+func (d *DACodecV7) NewDABlock(block *Block, totalL1MessagePoppedBefore uint64) (DABlock, error) {
+	if !block.Header.Number.IsUint64() {
+		return nil, errors.New("block number is not uint64")
+	}
+
+	numL1Messages, highestQueueIndex, err := block.NumL1MessagesNoSkipping()
+	if err != nil {
+		return nil, fmt.Errorf("failed to calculate number of L1 messages: %w", err)
+	}
+	if totalL1MessagePoppedBefore+uint64(numL1Messages) != highestQueueIndex {
+		return nil, fmt.Errorf("failed to sanity check L1 messages count: totalL1MessagePoppedBefore + numL1Messages != highestQueueIndex: %d + %d != %d", totalL1MessagePoppedBefore, numL1Messages, highestQueueIndex)
+	}
+
+	numL2Transactions := block.NumL2Transactions()
+	numTransactions := uint64(numL1Messages) + numL2Transactions
+	if numTransactions > math.MaxUint16 {
+		return nil, errors.New("number of transactions exceeds max uint16")
+	}
+
+	daBlock := newDABlockV7(
+		block.Header.Number.Uint64(), // number
+		block.Header.Time,            // timestamp
+		block.Header.BaseFee,         // baseFee
+		block.Header.GasLimit,        // gasLimit
+		uint16(numTransactions),      // numTransactions
+		numL1Messages,                // numL1Messages
+	)
+
+	return daBlock, nil
+}
+
+// NewDAChunk creates a new DAChunk from the given Chunk and the total number of L1 messages popped before.
+// Note: For DACodecV7, this function is not implemented since there is no notion of DAChunk in this version. Blobs
+// contain the entire batch data without any information of Chunks within.
+func (d *DACodecV7) NewDAChunk(_ *Chunk, _ uint64) (DAChunk, error) {
+	return nil, nil
+}
+
+// NewDABatch creates a DABatch including blob from the provided Batch.
+func (d *DACodecV7) NewDABatch(batch *Batch) (DABatch, error) {
+	if len(batch.Chunks) != 0 {
+		return nil, errors.New("batch must not contain any chunks")
+	}
+
+	if len(batch.Blocks) == 0 {
+		return nil, errors.New("batch must contain at least one block")
+	}
+
+	blob, blobVersionedHash, blobBytes, err := d.constructBlob(batch)
+	if err != nil {
+		return nil, fmt.Errorf("failed to construct blob: %w", err)
+	}
+
+	daBatch, err := newDABatchV7(CodecV7, batch.Index, blobVersionedHash, batch.ParentBatchHash, blob, blobBytes)
+	if err != nil {
+		return nil, fmt.Errorf("failed to construct DABatch: %w", err)
+	}
+
+	return daBatch, nil
+}
+
+func (d *DACodecV7) constructBlob(batch *Batch) (*kzg4844.Blob, common.Hash, []byte, error) {
+	enableCompression, err := d.CheckBatchCompressedDataCompatibility(batch)
```

Comment:
Just noting that it seems we're calling `constructBlobPayload` twice? Once inside `CheckBatchCompressedDataCompatibility` and once a couple of lines below. (But this is not introduced in this PR.)

Reply:
yep it was a remnant from how it worked before. You're right, I also don't see a reason to construct the blob twice and compress it twice. Changed in b71c047ffecc91720cd8110f9fac227c84ef3e29